Feature/core plugin loading logic

MrKarls opened the pull request.

MrKarls committed:
add loadPlugin function

MrKarls committed:
if checking mistake fixed

unitedtimur commented on core/src/core.cpp:
```diff
@@ -6,4 +6,69 @@ namespace Core {
+              }
+
+          }
+    }
```
Используй `.clang-format`, код не читаемый.

unitedtimur commented on core/include/core.h:
```diff
@@ -1,5 +1,12 @@
+#include "base_interface.h"
 namespace Core {
     void init();
+    void loadPlugins();
```
Надо сделать класс, не делать это в виде функций.

unitedtimur commented on core/include/core.h:
```diff
@@ -1,5 +1,12 @@
+#include <QMetaClassInfo>
+#include <QCoreApplication>
+#include "base_interface.h"
 namespace Core {
```
add `\n` above

unitedtimur commented on core/include/core.h:
```diff
@@ -1,5 +1,12 @@
+#include <QPluginLoader>
+#include <QMetaClassInfo>
+#include <QCoreApplication>
+#include "base_interface.h"
```
include зависимости текущего таргета должен быть выше всех.

unitedtimur commented on core/src/core.cpp:
```diff
@@ -6,4 +6,69 @@ namespace Core {
+        pluginsDir.cd("plugins");
+
+        if (!pluginsDir.exists())
+            throw std::runtime_error("Directories do not exist");
```
В qt не используют exceptions. Используй механизм вывода ошибки и отправки сигнала выхода из приложения.

unitedtimur commented on core/src/core.cpp:
```diff
@@ -6,4 +6,69 @@ namespace Core {
+              if(pluginObject)
+              {
+                 const QMetaObject *metaObject = pluginObject->metaObject();
+                 int index = metaObject->indexOfClassInfo("Dependencies");
```
А зачем так?

Reensef committed:
refactoring with .clang-format

Reensef committed:
add abstract_core.h

MrKarls committed:
add Core class,fix error processing

Reensef committed:
Merge branch 'feature/core-plugin-loading-logic' of https://github.com/unitedtimur/NotificationManager into feature/core-plugin-loading-logic

MrKarls committed:
class full refactoring

MrKarls committed:
Add class constructor

Reensef committed:
Merge branch 'feature/core-plugin-loading-logic' of https://github.com/unitedtimur/NotificationManager into feature/core-plugin-loading-logic

Reensef committed:
rename Core to AbstractCore and add namespace

Reensef committed:
fix naming

Reensef committed:
fix plugin loading

Reensef committed:
add link core to plugins

Reensef committed:
Merge branch 'master' into feature/core-plugin-loading-logic

bhaalgore committed:
ready to test

Reensef committed:
global fix

Reensef committed:
add class between base_interface and interface in plugin

Reensef committed:
change test output in test class

Reensef committed:
temp change

Reensef committed:
del extra files and finish gui plugin

Reensef committed:
add logic interface

Reensef committed:
add logic plugin and add namespaces

Reensef committed:
del description from project() in plugins cmake

Reensef committed:
del description from project() in plugins cmake

Reensef committed:
rm extra debug info in cmake

Reensef committed:
remake core

Reensef committed:
rm debug output in app and return in first state

Reensef committed:
Merge branch 'test/metadata-problem' into feature/core-plugin-loading-logic

Reensef committed:
add arg in initialize base_interface

Reensef committed:
rm mistake in IID gui plugin

Reensef committed:
rm plugin name

Reensef committed:
Merge branch 'master' into feature/core-plugin-loading-logic

Reensef committed:
set main app output directory

Reensef committed:
set plugins output directory

Reensef committed:
fix output directory in logic plugin

Reensef committed:
change plugin folder in core and main

Reensef committed:
rm unuseful commented code

Reensef committed:
add copying qml folder to install-root

I tried it in windows. Need to try in Linux

unitedtimur commented on core/include/abstract_core.h:
```diff
@@ -0,0 +1,45 @@
+#include <QSettings>
+#include <QQmlApplicationEngine>
+
+namespace NMCore {
```
Что за `NMCore`? Надо `Core`

unitedtimur commented on app/src/main.cpp:
```diff
@@ -10,9 +8,17 @@
+    QString plugDir = thisDir.absolutePath();
+
+    NMCore::AbstractCore core;
+    core.loadPlugins(plugDir);
```
Не надо вызывать так ядро приложения. Сделайте сущность приложения app, можно класс, а в нём вызывать `Core`

unitedtimur commented on core/include/abstract_core.h:
```diff
@@ -0,0 +1,45 @@
+
+        QPointer<QQmlApplicationEngine> qmlEngine() const;
+
+        // ...
```
?

unitedtimur commented on core/include/gui_interface.h:
```diff
@@ -3,22 +3,18 @@
 #include "base_interface.h"
 
-namespace Core {
+namespace NMCore {
```
Было норм.

unitedtimur commented on core/include/gui_interface.h:
```diff
@@ -3,22 +3,18 @@
 }
-Q_DECLARE_INTERFACE(Core::GuiInterface, "com.Core.GuiInterface")
+
+Q_DECLARE_INTERFACE(NMCore::GuiInterface, "com.NM.Core.GuiInterface")
```
Вот тут люто, верни как было.

unitedtimur commented on core/src/abstract_core.cpp:
```diff
@@ -0,0 +1,58 @@
+        //            }
+
+        //            return result;
+        //        };
```
?

unitedtimur commented on core/src/abstract_core.cpp:
```diff
@@ -0,0 +1,58 @@
+        for (const auto &plugin : qAsConst(plugins)) {
+            QPluginLoader loader(dir.absoluteFilePath(plugin));
+
+            QObject *obj = qobject_cast<QObject *>(loader.instance());
```
`const auto obj = ...`

unitedtimur commented on core/src/abstract_core.cpp:
```diff
@@ -0,0 +1,58 @@
+            QObject *obj = qobject_cast<QObject *>(loader.instance());
+
+            if (obj) {
+                qDebug() << "Plugin loaded";
```
`qDebug() << Q_FUNC_INFO <<  "Plugin loaded";`

unitedtimur commented on core/src/abstract_core.cpp:
```diff
@@ -0,0 +1,58 @@
+            if (obj) {
+                qDebug() << "Plugin loaded";
+            } else {
+                qWarning() << loader.errorString();
```
`qWarning() << Q_FUNC_INFO << loader.errorString();`

unitedtimur commented on plugins/gui/CMakeLists.txt:
```diff
@@ -44,6 +46,11 @@ target_include_directories(
+        ${PROJECT_NAME}
+        PRIVATE
+        ${PROJECT_NAMESPACE}::core
+)
```
Я бы изменил стиль, а то как то сложно читается.

```
target_link_libraries(${PROJECT_NAME}
        PRIVATE
        ${PROJECT_NAMESPACE}::core
)
```

unitedtimur commented on core/src/abstract_core.cpp:
```diff
@@ -0,0 +1,58 @@
+        QStringList plugins;
+        QDir dir(path);
+
+        if (dir.exists())
```
```
if (!dir.exists)
return;
```

unitedtimur commented on plugins/gui/include/gui_plugin.h:
```diff
@@ -0,0 +1,34 @@
+#include <QObject>
+#include <QString>
+#include <QTimer>
+namespace NMGuiPlugin {
```
add `\n` above

unitedtimur commented on plugins/gui/include/gui_plugin.h:
```diff
@@ -0,0 +1,34 @@
+         * используется при внедрении плагинов.
+         */
+        Q_INTERFACES(NMCore::GuiInterface)
+        /*!
```
add \n above

unitedtimur commented on plugins/gui/include/gui_plugin.h:
```diff
@@ -0,0 +1,34 @@
+    class GuiPlugin final : public NMCore::GuiInterface
+    {
+        Q_OBJECT
+        /*!
```
add \n above

unitedtimur commented on plugins/logic/CMakeLists.txt:
```diff
@@ -41,6 +45,11 @@ target_include_directories(
+        ${PROJECT_NAME}
+        PRIVATE
+        ${PROJECT_NAMESPACE}::core
+)
```
Аналогично

unitedtimur commented on plugins/logic/include/notify_logic_plugin.h:
```diff
@@ -5,26 +5,27 @@
+                */
+        Q_PLUGIN_METADATA(IID "com.NM.Core.LogicInterface" FILE "plugin_metadata.json")
 
     public:
```
Здесь тоже самое, включая стилистику файла.

unitedtimur commented on third_party/cmake/macro.cmake:
```diff
@@ -160,16 +160,8 @@ macro(PLUGIN_METADATA_GENERATOR NAME)
-    \"Description\": \"${_METADATA_DESCRIPTION}\",
-    \"Url\": \"${_METADATA_URL}\",
-    \"Dependencies\": \"${_METADATA_DEPENDENCIES}\"
+    \"Version\": \"${_METADATA_VERSION}\"
```
Так делать не надо. Используй генерацию из `.in`

unitedtimur replied:
https://cmake.org/cmake/help/latest/command/configure_file.html

unitedtimur commented on third_party/cmake/macro.cmake:
```diff
@@ -138,7 +138,7 @@ macro(GENERATE_EXPORT_HEADERS)
 
 # Макрос генерирует метадату для плагина
-macro(PLUGIN_METADATA_GENERATOR NAME)
+macro(PLUGIN_METADATA_GENERATOR TARGET)
```
Тогда уж `TARGET_NAME`

bhaalgore committed:
fix namespaces and class names, also plugins name in metadata file

bhaalgore committed:
fix source location

bhaalgore committed:
minor stylistics fix

bhaalgore committed:
additional stylistics fix

bhaalgore committed:
and more stylistics fixes

bhaalgore committed:
minor fix of include stylistics

bhaalgore committed:
Revert "add class PluginLoader"

This reverts commit 92626e1f86e74fb036ad8a42fa599685003c6042.

bhaalgore committed:
add class PluginLoader

bhaalgore committed:
fix wrong naming

bhaalgore committed:
add class PluginLoader

bhaalgore committed:
add CMake code

Добавил код в CMake лист, чтобы сущности внутри таргета app имели доступ к хидерам

bhaalgore committed:
relocate PluginLoader class

Reensef committed:
change way to generate metadata

unitedtimur commented on app/include/plugin_loader.h:
```diff
@@ -0,0 +1,30 @@
+     * \brief Класс, используемый для создания сущности, при помощи которой вызывается метод
+     * загрузки плагинов
+     */
+    class PluginLoader
```
Назови класс Application пж

unitedtimur commented on app/include/plugin_loader.h:
```diff
@@ -0,0 +1,30 @@
+    class PluginLoader
+    {
+    public:
+        PluginLoader();
```
`explicit`

unitedtimur commented on app/include/plugin_loader.h:
```diff
@@ -0,0 +1,30 @@
+
+    private:
+        Core::AbstractCore _core;
+        QString _pluginsDir;
```
Добавить описание

unitedtimur commented on app/include/plugin_loader.h:
```diff
@@ -0,0 +1,30 @@
+         * \brief Метод, вызывающий метод загрузки плагинов
+         * \param pluginsDir - переменная, указывающая на каталог с плагинами
+         */
+        void invokePluginsLoading(QDir pluginsDir);
```
`const QDir &`

unitedtimur commented on app/src/main.cpp:
```diff
@@ -10,9 +8,13 @@
+    thisDir.cdUp();
+    thisDir.cd("plugins");
+    App::PluginLoader pluginInvoker;
+    pluginInvoker.invokePluginsLoading(thisDir);
```
`QGuiApplication::applicationDirPath() + "/plugins/"`??????

bhaalgore replied:
тут не получится написать как ты в примере указал, потому что applicationDirPath возвращает путь до директории app, а plugins лежит на одном уровне с app, поэтому я обрезал у получившейся строки подстроку "/app", и после этого добавил к ней /plugins

unitedtimur commented on core/include/abstract_core.h:
```diff
@@ -0,0 +1,43 @@
+#include <QQmlApplicationEngine>
+
+namespace Core {
+    class CORE_EXPORT AbstractCore : public QObject
```
Описание добавить надо.

unitedtimur commented on core/include/abstract_core.h:
```diff
@@ -0,0 +1,43 @@
+
+        void loadPlugins(const QString &path);
+
+        QPointer<QQmlApplicationEngine> qmlEngine() const;
```
Здесь тоже.

unitedtimur commented on core/include/abstract_core.h:
```diff
@@ -0,0 +1,43 @@
+    public:
+        explicit AbstractCore();
+
+        void loadPlugins(const QString &path);
```
И здесь.

unitedtimur commented on core/include/base_interface.h:
```diff
@@ -4,27 +4,27 @@
-         * \brief Виртуальный метод, который реализует инициализацию плагинов логики и gui, исходя
-         * из зависимостей библиотек друг от друга
+         * \brief Виртуальный метод, который реализует инициализацию плагинов
+         * логики и gui, исходя из зависимостей библиотек друг от друга
```
Описание некорректное. 

Метод позволяет получать список всех динамических библиотек

unitedtimur commented on core/include/base_interface.h:
```diff
@@ -4,27 +4,27 @@
     class BaseInterface : public QObject
     {
         Q_OBJECT
     public:
```
add `\n` above

unitedtimur commented on core/include/logic_interface.h:
```diff
@@ -16,6 +16,7 @@ namespace Core {
 }
-Q_DECLARE_INTERFACE(Core::LogicInterface, "com.Core.LogicInterface")
+
+Q_DECLARE_INTERFACE(Core::LogicInterface, "com.NM.Core.LogicInterface")
```
Смотри что я писал уже.

unitedtimur commented on plugins/gui/include/notification_gui_plugin.h:
```diff
@@ -1,34 +1,37 @@
+        Q_PLUGIN_METADATA(IID "com.NotificationGuiPlugin" FILE "plugin_metadata.json")
+
+    public:
+        virtual bool initialize(const QList<QPointer<QObject>> &dependencies) override;
```
Зачем виртуальный метод сделал здесь?

unitedtimur commented on plugins/gui/include/notification_gui_plugin.h:
```diff
@@ -1,34 +1,37 @@
+         * интерфейса, реализованного через объект, и сослаться на файл, содержащий метаданные для
+         * плагина.
+         */
+        Q_PLUGIN_METADATA(IID "com.NotificationGuiPlugin" FILE "plugin_metadata.json")
```
Можно файл назвать просто `metadata.json`?

unitedtimur commented on plugins/logic/include/notification_logic_plugin.h:
```diff
@@ -0,0 +1,35 @@
+        Q_PLUGIN_METADATA(IID "com.Core.LogicInterface" FILE "plugin_metadata.json")
+
+    public:
+        virtual bool initialize(const QList<QPointer<QObject>> &dependencies) override;
```
Тут опять всё поехало.

unitedtimur commented on third_party/cmake/macro.cmake:
```diff
@@ -138,38 +138,47 @@ macro(GENERATE_EXPORT_HEADERS)
 
 # Макрос генерирует метадату для плагина
-macro(PLUGIN_METADATA_GENERATOR NAME)
+# Используйте слудующий синтаксис
```
В описании никто не надо использовать какое-то первое лицо, надо говорить что делает тот или иной блок кода.

bhaalgore committed:
rename pluginLoader class, delete member _pluginsDir due other fix

bhaalgore committed:
fixing a variable pointing to a plugins dir

unitedtimur commented on app/src/main.cpp:
```diff
@@ -9,12 +9,12 @@
+                          QGuiApplication::applicationDirPath().lastIndexOf('/'))
+                         + "/plugins"; // тут не получится написать как ты в примере указал, потому
+                                       // что applicationDirPath возвращает путь до директории app,
+                                       // а plugins лежит на одном уровне с app
```
Что? Во-первых зачем комментарий адресованный кому-то, тут код пишется как бы командой.

Второе - всё код делает верно, который кидал ранее, просто нужно с умом подойти к процессу.
У тебя бинарник лежит на уровне с папкой `plugins`. А код выше это что-то невероятное.

bhaalgore committed:
add description of method

bhaalgore committed:
delete unnecessary comments, fix description and interface IID naming

bhaalgore committed:
fix IID naming, rem unnecessary specificator virtual

bhaalgore committed:
Merge branch 'feature/install-root' into feature/core-plugin-loading-logic

bhaalgore committed:
change way to load plugins

bhaalgore committed:
add spaces and fix descriptions

bhaalgore committed:
change metadata naming, rem uncorrect coments in CMake

bhaalgore committed:
fix plugins loading

bhaalgore committed:
a lot of stylistics fixes

Reensef committed:
add description to abstract core

Reensef committed:
change description to abstract core

bhaalgore committed:
small fix
